crates/biome_js_formatter/src/js/expressions/template_expression.rs

```diff
@@ -99,3 +138,333 @@ impl NeedsParentheses for JsTemplateExpression {
+                            element.clone(),
+                            placeholder_map,
+                            f,
+                        )
+                    })
+                    .collect::<Result<Vec<FormatElement>, _>>()?;
+                Ok(FormatElement::Interned(Interned::new(elemets)))
+            }
+            FormatElement::BestFitting(best_fitting) => {
+                let variants = best_fitting
+                    .variants()
+                    .iter()
+                    .map(|variant| {
+                        let elements = variant
+                            .iter()
+                            .map(|element| {
+                                replace_placeholder_with_template_element(
+                                    element.clone(),
+                                    placeholder_map,
+                                    f,
+                                )
+                            })
+                            .collect::<Result<Vec<FormatElement>, _>>();
+                        elements.map(|elements| Box::new(elements).into_boxed_slice())
+                    })
+                    .collect::<Result<Vec<Box<[FormatElement]>>, _>>()?;
+
+                // SAFETY: the best fitting element is derived from a valid best fitting element
+                unsafe {
+                    Ok(FormatElement::BestFitting(
+                        BestFittingElement::from_vec_unchecked(variants),
+                    ))
+                }
+            }
+            element => Ok(element),
+        }
+    }
+
+    let mut iter = embedded_language_formatted.iter().peekable();
+    let mut formatted: Vec<FormatElement> = Vec::new();
+    while let Some(element) = iter.next() {
+        if let FormatElement::LocatedTokenText { slice, .. } = element {
+            let text = slice.to_string();
+            if text == "$" {
+                if let Some(FormatElement::LocatedTokenText { slice, .. }) = iter.peek() {
+                    let next_text = slice.to_string();
+                    if placeholder_map.contains_key(&next_text) {
+                        continue;
+                    }
+                }
+            }
+        }
+        formatted.push(replace_placeholder_with_template_element(
+            element.clone(),
+            &mut placeholder_map,
+            f,
+        )?);
+    }
+    // replace placeholders with the original string interpolations
+    // let formatted = embedded_language_formatted
+    //     .deref()
+    //     .iter()
+    //     .map(|element| {
+    //         replace_placeholder_with_template_element(element.clone(), &mut placeholder_map, f)
+    //     })
+    //     .collect::<Result<Vec<FormatElement>, _>>()?;
+
+    // if there are any placeholders left, we treat it is a error and format the template as normal
+    if !placeholder_map.is_empty() {
+        return Err(FormatError::SyntaxError);
+    }
```

Comment:
I would envision use cases like the one below will be reasonable and may be common ones:

```js
const index = 1;

const code = css`
div {
  background-image: url("./assets/images/img-${index}.png");
}`;
```

This is a sub-node level interpolation, I'm not sure whether the current `replace_placeholder_with_template_element` will recover such interpolations properly? I see currently we bail out if we don't restore all the interpolations from the placeholders, and format this as a normal js template string, which is good. But I guess we can keep this in mind and consider possible solutions? 🤔

Reply:
> I'm not sure whether the current replace_placeholder_with_template_element will recover such interpolations properly?

Unfortunately no.

Currently, this method does an exact text match with placeholders. However, the code snippet you provided will produce a format element like `text("./assets/images/img-$placeholder.png")`. Maybe we can implement partial matching with placeholders and split the element into multiple segments, something like `[text("./assets/images/img-"), expr, text(".png")]`.

Reply:
> Maybe we can implement partial matching with placeholders and split the element into multiple segments, something like `[text("./assets/images/img-"), expr, text(".png")]`

👍 Yes, I think we don't need an exact match.